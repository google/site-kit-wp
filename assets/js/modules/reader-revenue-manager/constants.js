/**
 * Reader Revenue Manager constants.
 *
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const SNIPPET_MODES = {
	post_types: __( 'Specific content types', 'google-site-kit' ),
	per_post: __( 'Specified pages', 'google-site-kit' ),
	sitewide: __( 'Site wide', 'google-site-kit' ),
};

export const RRM_SETUP_NOTIFICATION_ID = 'rrm-setup-notification';
export const RRM_SETUP_SUCCESS_NOTIFICATION_ID =
	'setup-success-notification-rrm';
export const RRM_PRODUCT_ID_CONTRIBUTIONS_NOTIFICATION_ID =
	'rrm-product-id-contributions-notification';
export const RRM_PRODUCT_ID_SUBSCRIPTIONS_NOTIFICATION_ID =
	'rrm-product-id-subscriptions-notification';
