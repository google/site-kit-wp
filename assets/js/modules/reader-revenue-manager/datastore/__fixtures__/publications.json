[
	{
		"publicationId": "ABCDEFGH",
		"publicationPredicates": {
			"businessPredicates": {
				"supportsSiteKit": true,
				"canSell": true
			}
		},
		"verifiedDomains": "example.com",
		"paymentOptions": {
			"subscriptions": true,
			"noPayment": false,
			"contributions": false,
			"thankStickers": true
		},
		"displayName": "Test Publication 1",
		"products": [
			{
				"name": "basic"
			}
		],
		"onboardingState": "ONBOARDING_STATE_UNSPECIFIED"
	},
	{
		"publicationId": "PQRSTUV",
		"publicationPredicates": {
			"businessPredicates": {
				"supportsSiteKit": true,
				"canSell": false
			}
		},
		"verifiedDomains": "example.com",
		"paymentOptions": {
			"subscriptions": true,
			"noPayment": false,
			"contributions": false,
			"thankStickers": true
		},
		"displayName": "Test Publication 2",
		"products": [
			{
				"name": "premium"
			}
		],
		"onboardingState": "PENDING_VERIFICATION"
	},
	{
		"publicationId": "IJKLMNOP",
		"publicationPredicates": {
			"businessPredicates": {
				"supportsSiteKit": true,
				"canSell": true
			}
		},
		"verifiedDomains": "example.com",
		"paymentOptions": {
			"subscriptions": true,
			"noPayment": false,
			"contributions": false,
			"thankStickers": true
		},
		"displayName": "Test Publication 3",
		"products": [
			{
				"name": "basic"
			}
		],
		"onboardingState": "ONBOARDING_ACTION_REQUIRED"
	},
	{
		"publicationId": "QRSTUVWX",
		"publicationPredicates": {
			"businessPredicates": {
				"supportsSiteKit": false,
				"canSell": true
			}
		},
		"verifiedDomains": "example.com",
		"paymentOptions": {
			"subscriptions": true,
			"noPayment": false,
			"contributions": true,
			"thankStickers": true
		},
		"displayName": "Test Publication 4",
		"products": [
			{
				"name": "premium"
			}
		],
		"onboardingState": "ONBOARDING_COMPLETE"
	}
]
