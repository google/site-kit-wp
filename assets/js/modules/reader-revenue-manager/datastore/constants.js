/**
 * `modules/reader-revenue-manager` data store constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const ERROR_CODE_NON_HTTPS_SITE = 'non_https_site';

export const MODULES_READER_REVENUE_MANAGER = 'modules/reader-revenue-manager';

export const READER_REVENUE_MANAGER_MODULE_SLUG = 'reader-revenue-manager';

export const PUBLICATION_ONBOARDING_STATES = {
	ONBOARDING_COMPLETE: 'ONBOARDING_COMPLETE',
	ONBOARDING_ACTION_REQUIRED: 'ONBOARDING_ACTION_REQUIRED',
	PENDING_VERIFICATION: 'PENDING_VERIFICATION',
	UNSPECIFIED: 'ONBOARDING_STATE_UNSPECIFIED',
};

export const UI_KEY_READER_REVENUE_MANAGER_SHOW_PUBLICATION_APPROVED_NOTIFICATION =
	'READER_REVENUE_MANAGER_SHOW_PUBLICATION_APPROVED_NOTIFICATION';

export const LEGACY_RRM_SETUP_BANNER_DISMISSED_KEY =
	'rrm_module_setup_banner_dismissed_key';

export const READER_REVENUE_MANAGER_SETUP_FORM =
	'readerRevenueManagerSetupForm';

export const READER_REVENUE_MANAGER_NOTICES_FORM =
	'readerRevenueManagerNoticesForm';

export const SHOW_PUBLICATION_CREATE = 'showPublicationCreate';

export const RESET_PUBLICATIONS = 'resetPublications';

export const SYNC_PUBLICATION = 'syncPublication';
