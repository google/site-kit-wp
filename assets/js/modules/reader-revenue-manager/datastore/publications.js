/**
 * `modules/reader-revenue-manager` data store: publications.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';
import { isPlainObject } from 'lodash';

/**
 * Internal dependencies.
 */
import API from 'googlesitekit-api';
import {
	commonActions,
	combineStores,
	createReducer,
} from 'googlesitekit-data';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { createValidatedAction } from '../../../googlesitekit/data/utils';
import {
	MODULES_READER_REVENUE_MANAGER,
	READER_REVENUE_MANAGER_MODULE_SLUG,
	PUBLICATION_ONBOARDING_STATES,
} from './constants';
import { actions as errorStoreActions } from '../../../googlesitekit/data/create-error-store';

const fetchGetPublicationsStore = createFetchStore( {
	baseName: 'getPublications',
	controlCallback: () =>
		API.get(
			'modules',
			READER_REVENUE_MANAGER_MODULE_SLUG,
			'publications',
			{},
			{ useCache: false }
		),
	reducerCallback: ( state, publications ) => ( { ...state, publications } ),
} );

const fetchGetSyncPublicationOnboardingStateStore = createFetchStore( {
	baseName: 'getSyncPublicationOnboardingState',
	controlCallback: ( { publicationID, publicationOnboardingState } ) =>
		API.set(
			'modules',
			READER_REVENUE_MANAGER_MODULE_SLUG,
			'sync-publication-onboarding-state',
			{
				publicationID,
				publicationOnboardingState,
			}
		),
	argsToParams: ( { publicationID, publicationOnboardingState } ) => {
		return { publicationID, publicationOnboardingState };
	},
	validateParams: ( { publicationID, publicationOnboardingState } = {} ) => {
		invariant(
			typeof publicationID === 'string' && publicationID.length > 0,
			'publicationID is required and must be string.'
		);

		invariant(
			typeof publicationOnboardingState === 'string' &&
				publicationOnboardingState.length > 0,
			'publicationOnboardingState is required and must be string.'
		);
	},
	reducerCallback: createReducer(
		( state, { publicationID, publicationOnboardingState } ) => {
			if ( ! publicationID ) {
				return;
			}

			// eslint-disable-next-line sitekit/no-direct-date
			const publicationOnboardingStateLastSyncedAtMs = Date.now();

			if ( state.settings.publicationID === publicationID ) {
				state.settings.publicationOnboardingState =
					publicationOnboardingState;
				state.settings.publicationOnboardingStateLastSyncedAtMs =
					publicationOnboardingStateLastSyncedAtMs;
			}

			if ( state.savedSettings.publicationID === publicationID ) {
				state.savedSettings.publicationOnboardingState =
					publicationOnboardingState;
				state.savedSettings.publicationOnboardingStateLastSyncedAtMs =
					publicationOnboardingStateLastSyncedAtMs;
			}

			const publication = state.publications?.find(
				// eslint-disable-next-line sitekit/acronym-case
				( { publicationId: id } ) => id === publicationID
			);

			if ( publication ) {
				publication.onboardingState = publicationOnboardingState;
			}
		}
	),
} );

const baseInitialState = {
	publications: undefined,
};

const baseActions = {
	/**
	 * Synchronizes the onboarding state of the publication with the API.
	 * Updates the settings on the server.
	 *
	 * @since 1.132.0
	 *
	 * @return {void}
	 */
	*syncPublicationOnboardingState() {
		const registry = yield commonActions.getRegistry();

		yield commonActions.await(
			registry
				.resolveSelect( MODULES_READER_REVENUE_MANAGER )
				.getSettings()
		);

		const publicationID = registry
			.select( MODULES_READER_REVENUE_MANAGER )
			.getPublicationID();

		const publicationOnboardingState = registry
			.select( MODULES_READER_REVENUE_MANAGER )
			.getPublicationOnboardingState();

		// If there is no publication ID in state, do not attempt to sync
		// the onboarding state.
		if (
			publicationID === undefined ||
			publicationOnboardingState === undefined
		) {
			return {};
		}

		return yield fetchGetSyncPublicationOnboardingStateStore.actions.fetchGetSyncPublicationOnboardingState(
			{
				publicationID,
				publicationOnboardingState,
			}
		);
	},

	/**
	 * Finds a matched publication.
	 *
	 * @since 1.132.0
	 *
	 * @return {Object|null} Matched publication; `null` if none found.
	 */
	*findMatchedPublication() {
		const { resolveSelect } = yield commonActions.getRegistry();
		const publications = yield commonActions.await(
			resolveSelect( MODULES_READER_REVENUE_MANAGER ).getPublications()
		);

		if ( publications.length === 0 ) {
			return null;
		}

		if ( publications.length === 1 ) {
			return publications[ 0 ];
		}

		const completedOnboardingPublication = publications.find(
			( publication ) =>
				publication.onboardingState ===
				PUBLICATION_ONBOARDING_STATES.ONBOARDING_COMPLETE
		);

		return completedOnboardingPublication || publications[ 0 ];
	},

	/**
	 * Resets the publications data in the store.
	 *
	 * @since 1.133.0
	 *
	 * @return {Object} The dispatched action results.
	 */
	*resetPublications() {
		const registry = yield commonActions.getRegistry();

		yield {
			type: 'RESET_PUBLICATIONS',
		};

		yield errorStoreActions.clearErrors( 'getPublications' );

		return registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.invalidateResolutionForStoreSelector( 'getPublications' );
	},

	/**
	 * Sets the given publication in the store.
	 *
	 * @since 1.133.0
	 *
	 * @param {Object} publication The publiation object.
	 * @return {Object} A Generator function.
	 */
	selectPublication: createValidatedAction(
		( publication ) => {
			invariant(
				isPlainObject( publication ),
				'A valid publication object is required.'
			);

			[ 'publicationId', 'onboardingState' ].forEach( ( key ) => {
				invariant(
					publication.hasOwnProperty( key ),
					`The publication object must contain ${ key }`
				);
			} );
		},
		// `publicationId` is the identifier used by the API.
		// eslint-disable-next-line sitekit/acronym-case
		function* ( { publicationId: publicationID, onboardingState } ) {
			const registry = yield commonActions.getRegistry();

			return registry
				.dispatch( MODULES_READER_REVENUE_MANAGER )
				.setSettings( {
					publicationID,
					publicationOnboardingState: onboardingState,
				} );
		}
	),
};

const baseControls = {};

const baseReducer = ( state, { type } ) => {
	switch ( type ) {
		case 'RESET_PUBLICATIONS':
			return {
				...state,
				publications: baseInitialState.publications,
			};
		default:
			return state;
	}
};

const baseResolvers = {
	*getPublications() {
		const registry = yield commonActions.getRegistry();
		// Only fetch publications if there are none in the store.
		const publications = registry
			.select( MODULES_READER_REVENUE_MANAGER )
			.getPublications();
		if ( publications === undefined ) {
			yield fetchGetPublicationsStore.actions.fetchGetPublications();
		}
	},
};

const baseSelectors = {
	/**
	 * Gets list of publications associated with the account.
	 *
	 * @since 1.132.0
	 *
	 * @param {Object} state Data store's state.
	 * @return {(Array.<Object>|undefined)} An array of publications; `undefined` if not loaded.
	 */
	getPublications( state ) {
		return state.publications;
	},
};

const store = combineStores(
	fetchGetPublicationsStore,
	fetchGetSyncPublicationOnboardingStateStore,
	{
		initialState: baseInitialState,
		actions: baseActions,
		controls: baseControls,
		reducer: baseReducer,
		resolvers: baseResolvers,
		selectors: baseSelectors,
	}
);

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
