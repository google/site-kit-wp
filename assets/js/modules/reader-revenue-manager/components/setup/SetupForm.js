/**
 * Reader Revenue Manager Setup form.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { useCallback, useEffect } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { SpinnerButton } from 'googlesitekit-components';
import { useDispatch, useSelect } from 'googlesitekit-data';
import Link from '../../../../components/Link';
import StoreErrorNotices from '../../../../components/StoreErrorNotices';
import { CORE_FORMS } from '../../../../googlesitekit/datastore/forms/constants';
import {
	READER_REVENUE_MANAGER_MODULE_SLUG,
	MODULES_READER_REVENUE_MANAGER,
	READER_REVENUE_MANAGER_SETUP_FORM,
	RESET_PUBLICATIONS,
} from '../../datastore/constants';
import { PublicationOnboardingStateNotice, PublicationSelect } from '../common';

export default function SetupForm( { onCompleteSetup } ) {
	const canSubmitChanges = useSelect( ( select ) =>
		select( MODULES_READER_REVENUE_MANAGER ).canSubmitChanges()
	);
	const isSaving = useSelect( ( select ) =>
		select( MODULES_READER_REVENUE_MANAGER ).isDoingSubmitChanges()
	);
	const publications = useSelect( ( select ) =>
		select( MODULES_READER_REVENUE_MANAGER ).getPublications()
	);
	const publicationID = useSelect( ( select ) =>
		select( MODULES_READER_REVENUE_MANAGER ).getPublicationID()
	);
	const createPublicationURL = useSelect( ( select ) =>
		select( MODULES_READER_REVENUE_MANAGER ).getCreatePublicationLinkURL()
	);

	const { setValues } = useDispatch( CORE_FORMS );
	const { findMatchedPublication, selectPublication } = useDispatch(
		MODULES_READER_REVENUE_MANAGER
	);

	const handleLinkClick = useCallback( () => {
		setValues( READER_REVENUE_MANAGER_SETUP_FORM, {
			[ RESET_PUBLICATIONS ]: true,
		} );
	}, [ setValues ] );

	const submitForm = useCallback(
		( event ) => {
			event.preventDefault();
			onCompleteSetup();
		},
		[ onCompleteSetup ]
	);

	// Automatically pre-select a publication.
	useEffect( () => {
		const autoSelectPublication = async () => {
			const matchedPublication = await findMatchedPublication();

			if ( matchedPublication ) {
				selectPublication( matchedPublication );
			}
		};

		if ( ! publicationID ) {
			autoSelectPublication();
		}
	}, [ findMatchedPublication, publicationID, selectPublication ] );

	if ( ! publications ) {
		return null;
	}

	return (
		<form onSubmit={ submitForm }>
			<StoreErrorNotices
				moduleSlug={ READER_REVENUE_MANAGER_MODULE_SLUG }
				storeName={ MODULES_READER_REVENUE_MANAGER }
			/>
			<p className="googlesitekit-margin-bottom-0">
				{ publications.length === 1
					? __(
							'Site Kit will connect your existing publication',
							'google-site-kit'
					  )
					: __(
							'Select your preferred publication to connect with Site Kit',
							'google-site-kit'
					  ) }
			</p>
			<div className="googlesitekit-setup-module__inputs">
				<PublicationSelect />
			</div>
			<PublicationOnboardingStateNotice />
			<Link
				external
				href={ createPublicationURL }
				onClick={ handleLinkClick }
			>
				{ __( 'Create new publication', 'google-site-kit' ) }
			</Link>
			<div className="googlesitekit-setup-module__action">
				<SpinnerButton
					disabled={ ! canSubmitChanges || isSaving }
					isSaving={ isSaving }
				>
					{ __( 'Complete setup', 'google-site-kit' ) }
				</SpinnerButton>
			</div>
		</form>
	);
}

SetupForm.propTypes = {
	onCompleteSetup: PropTypes.func.isRequired,
};
