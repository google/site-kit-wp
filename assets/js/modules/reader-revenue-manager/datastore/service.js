/**
 * `modules/reader-revenue-manager` data store: service.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { addQueryArgs } from '@wordpress/url';

/**
 * Internal dependencies
 */
import { createRegistrySelector } from 'googlesitekit-data';
import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';

const selectors = {
	/**
	 * Returns a link to the Reader Revenue Manager platform.
	 *
	 * @since 1.132.0
	 * @since 1.136.0 Removed `publicationID` arg.
	 *
	 * @param {Object} state                Data store's state.
	 * @param {Object} [args]               Object containing optional publication ID, path and query args.
	 * @param {string} [args.publicationID] Publication ID to be used in the URL.
	 * @param {string} [args.path]          A path to append to the base url.
	 * @param {Object} [args.query]         Object of query params to be added to the URL.
	 * @return {(string|undefined)} The URL to the service, or `undefined` if not loaded.
	 */
	getServiceURL: createRegistrySelector(
		( select ) =>
			( state, { path, query } = {} ) => {
				let serviceURL = 'https://publishercenter.google.com';

				if ( path ) {
					const sanitizedPath = `/${ path.replace( /^\//, '' ) }`;
					serviceURL = `${ serviceURL }${ sanitizedPath }`;
				}

				serviceURL = addQueryArgs( serviceURL, {
					...query,
					utm_source: 'sitekit', // Always add the utm_source.
				} );

				const accountChooserBaseURI =
					select( CORE_USER ).getAccountChooserURL( serviceURL );

				if ( accountChooserBaseURI === undefined ) {
					return undefined;
				}

				return accountChooserBaseURI;
			}
	),
};

const store = {
	selectors,
};

export default store;
