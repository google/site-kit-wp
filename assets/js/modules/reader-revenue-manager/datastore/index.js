/**
 * `modules/reader-revenue-manager` data store.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { combineStores } from 'googlesitekit-data';
import { MODULES_READER_REVENUE_MANAGER } from './constants';
import baseModuleStore from './base';
import publications from './publications';
import service from './service';

const store = combineStores( baseModuleStore, publications, service );

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export const registerStore = ( registry ) => {
	registry.registerStore( MODULES_READER_REVENUE_MANAGER, store );
};

export default store;
