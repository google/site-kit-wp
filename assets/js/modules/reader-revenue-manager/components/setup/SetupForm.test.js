/**
 * Reader Revenue Manager SetupForm component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { MODULES_READER_REVENUE_MANAGER } from '../../datastore/constants';
import {
	createTestRegistry,
	provideModules,
} from '../../../../../../tests/js/utils';
import {
	act,
	fireEvent,
	render,
	waitFor,
} from '../../../../../../tests/js/test-utils';
import { publications } from '../../datastore/__fixtures__';
import SetupForm from './SetupForm';

describe( 'SetupForm', () => {
	let registry;

	beforeEach( () => {
		registry = createTestRegistry();

		provideModules( registry );

		registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.receiveGetSettings( {} );
	} );

	it( 'should render the form correctly', async () => {
		registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.receiveGetPublications( publications );

		const { container, getByRole, getByText, waitForRegistry } = render(
			<SetupForm onCompleteSetup={ () => {} } />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( container ).toMatchSnapshot();

		expect( getByText( 'Publication' ) ).toBeInTheDocument();
		expect(
			getByRole( 'button', { name: /create new publication/i } )
		).toBeInTheDocument();
	} );

	it( 'should change instruction text based on number of publications', async () => {
		registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.receiveGetPublications( [ publications[ 0 ] ] );

		const { getByText, waitForRegistry } = render(
			<SetupForm onCompleteSetup={ () => {} } />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect(
			getByText( 'Site Kit will connect your existing publication' )
		).toBeInTheDocument();

		await act( () =>
			registry
				.dispatch( MODULES_READER_REVENUE_MANAGER )
				.receiveGetPublications( publications )
		);

		await waitForRegistry();

		expect(
			getByText(
				'Select your preferred publication to connect with Site Kit'
			)
		).toBeInTheDocument();
	} );

	it( 'should automatically find and select a matched publication', async () => {
		registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.receiveGetPublications( publications );

		const matchedPublication = await registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.findMatchedPublication();

		expect(
			registry.select( MODULES_READER_REVENUE_MANAGER ).getPublicationID()
		).toBeUndefined();

		const { waitForRegistry } = render(
			<SetupForm onCompleteSetup={ () => {} } />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect(
			registry.select( MODULES_READER_REVENUE_MANAGER ).getPublicationID()
			// eslint-disable-next-line sitekit/acronym-case
		).toEqual( matchedPublication.publicationId );
	} );

	it( 'should submit the form upon pressing the CTA', async () => {
		registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.receiveGetPublications( publications );

		const onCompleteSetup = jest.fn();

		const { getByRole, waitForRegistry } = render(
			<SetupForm onCompleteSetup={ onCompleteSetup } />,
			{
				registry,
			}
		);

		await waitForRegistry();

		act( () => {
			fireEvent.click(
				getByRole( 'button', { name: /Complete setup/i } )
			);
		} );

		await waitFor( () => {
			expect( onCompleteSetup ).toHaveBeenCalledTimes( 1 );
		} );
	} );
} );
