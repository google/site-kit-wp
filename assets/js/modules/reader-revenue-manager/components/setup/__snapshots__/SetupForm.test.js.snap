// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SetupForm > should render the form correctly 1`] = `
<div>
  <form>
    <p
      class="googlesitekit-margin-bottom-0"
    >
      Select your preferred publication to connect with Site Kit
    </p>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="mdc-select mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="QRSTUVWX"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Test Publication 4 (QRSTUVWX)
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-0"
              style=""
            >
              Publication
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="premium"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          premium
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-1"
              style=""
            >
              Default Product ID
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
    </div>
    <button
      class="googlesitekit-cta-link"
    >
      <span
        class="googlesitekit-cta-link__contents"
      >
        Manage publications in Publisher Center
      </span>
      <span
        class="googlesitekit-icon-wrapper"
        style="margin-left: 5px;"
      >
        <svg />
      </span>
    </button>
    <div
      class="googlesitekit-setup-module__action"
    >
      <button
        class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Complete setup
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`SetupForm should render the form correctly 1`] = `
<div>
  <form>
    <p
      class="googlesitekit-margin-bottom-0"
    >
      Select your preferred publication to connect with Site Kit
    </p>
    <div
      class="googlesitekit-setup-module__inputs"
    >
      <div
        class="mdc-select mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="QRSTUVWX"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          Test Publication 4 (QRSTUVWX)
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-0"
              style=""
            >
              Publication
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
      <div
        class="mdc-select mdc-select--outlined"
      >
        <i
          class="mdc-select__dropdown-icon"
        />
        <input
          name="enhanced-select"
          type="hidden"
          value="premium"
        />
        <div
          aria-disabled="false"
          class="mdc-select__selected-text"
          tabindex="0"
        >
          premium
        </div>
        <div
          class="mdc-notched-outline mdc-notched-outline--upgraded mdc-notched-outline--notched"
        >
          <div
            class="mdc-notched-outline__leading"
          />
          <div
            class="mdc-notched-outline__notch"
          >
            <label
              class="mdc-floating-label mdc-floating-label--float-above"
              for="googlesitekit-select-1"
              style=""
            >
              Default Product ID
            </label>
          </div>
          <div
            class="mdc-notched-outline__trailing"
          />
        </div>
      </div>
    </div>
    <button
      class="googlesitekit-cta-link"
    >
      <span
        class="googlesitekit-cta-link__contents"
      >
        Manage publications in Publisher Center
      </span>
      <span
        class="googlesitekit-icon-wrapper"
        style="margin-left: 5px;"
      >
        <svg />
      </span>
    </button>
    <div
      class="googlesitekit-setup-module__action"
    >
      <button
        class="mdc-button googlesitekit-button-icon--spinner googlesitekit-button-icon--spinner__after mdc-button--raised"
        target="_self"
      >
        <span
          class="mdc-button__label"
        >
          Complete setup
        </span>
      </button>
    </div>
  </form>
</div>
`;
